cpp/src/arrow/dataset/file_ipc.cc

```diff
@@ -57,22 +58,40 @@ class IpcScanTask : public ScanTask {
       : ScanTask(std::move(options), std::move(context)), source_(std::move(source)) {}
 
   Result<RecordBatchIterator> Execute() override {
-    struct {
+    struct Impl {
+      static Result<Impl> Make(const FileSource& source,
+                               const std::vector<std::string>& materialized_fields,
+                               MemoryPool* pool) {
+        ARROW_ASSIGN_OR_RAISE(auto reader, OpenReader(source));
+        auto materialized_schema =
+            SchemaFromColumnNames(reader->schema(), materialized_fields);
+        return Impl{std::move(reader),
+                    RecordBatchProjector(std::move(materialized_schema)), pool, 0};
+      }
+
       Result<std::shared_ptr<RecordBatch>> Next() {
         if (i_ == reader_->num_record_batches()) {
           return nullptr;
         }
 
         std::shared_ptr<RecordBatch> batch;
         RETURN_NOT_OK(reader_->ReadRecordBatch(i_++, &batch));
-        return batch;
+        return projector_.Project(*batch, pool_);
       }
 
       std::shared_ptr<ipc::RecordBatchFileReader> reader_;
-      int i_ = 0;
-    } batch_it;
-
-    ARROW_ASSIGN_OR_RAISE(batch_it.reader_, OpenReader(source_));
+      RecordBatchProjector projector_;
+      MemoryPool* pool_;
+      int i_;
+    };
+
+    // get names of fields explicitly projected or referenced by filter
+    auto fields = options_->MaterializedFields();
+    std::sort(fields.begin(), fields.end());
```

Comment:
We shouldn't do this, it will break when we start support columns with the same name but different index. I'd leave it as-is from `MaterializedFields`.

Reply:
@fsaintjacques since a non trivial filter will reference a field in the fragment's physical schema, that will result in a schema with duplicate field names. That breaks filter evaluation and other parts of our code which rely on `Schema::GetFieldIndex` (which returns "not found" if the name is repeated). I'll reintroduce the unique-ing and we can address this more robustly in https://issues.apache.org/jira/browse/ARROW-7412

Reply:
Indeed, ok for now!